### ISSUE TITLE
Opacity flashing fix introduced chicken and egg issue with opacity inheritance

### ISSUE THREAD

**Issue by flar** [flutter/flutter#121740]
The fix for https://github.com/flutter/flutter/issues/120455, introduced in https://github.com/flutter/engine/pull/39690 caused a shift in some of the memory benchmarks that measure cache usage from picture caching to layer caching.

This wasn't predicted, but is to be expected due to the nature of the fix. Without the fix we get noticeable flashing so reverting the fix isn't appropriate and the memory merely shifted rather than increased and the time metrics didn't flag a regression. So, there is no big practical measurable regression from this issue.

Picture metrics decreasing:
https://flutter-flutter-perf.skia.org/e/?begin=1676609568&end=1676681408&keys=X91060e41b7e82bf83ef90d3fe5f84a35&num_commits=50&request_type=1&xbaroffset=33419

Layer metrics increasing:
https://flutter-flutter-perf.skia.org/e/?begin=1676609568&end=1676681408&keys=X854efad91537cb086050237bfad88626&num_commits=50&request_type=1&xbaroffset=33419

The root cause is that picture layers used to declare themselves as able to inherit opacity when they became eligible to cache. The problem is that "eligible to cache" doesn't mean that they will be successful. In particular, we only allow 3 pictures to cache per frame so if there were more than 3 children than all of them might say they can inherit opacity, but only 3 would be telling the truth. The fix requires the children to wait until they successfully cache themselves before they declare that they can inherit opacity. The problem is that if the parent is caching the children then it will tell the children not to cache themselves. This policy prevents double-caching where the parent is caching the children who are also caching themselves. Unfortunately, since the children can never cache themselves, they also never declare that they can inherit opacity. Thus, a chicken and egg situation.

The fix will require some careful decisions about the eligibility and potential eligibility of the children to cache themselves and thus inherit opacity.

**Comment by flar** [comment]
CC @JsouLiang @ColdPaleLight I'm not sure there is a good solution to this issue that doesn't involve prescience as to whether all of the children might successfully cache themselves.

**Comment by JsouLiang** [comment]
The DL can be cached and cached successful so that it can be inherit opacity.
If not this, the OpacityLayer may be use `saveLayer` op. I don't know if we could connect this issue with that `LayerStack` solution?
Another a issue **provide a more transparent mechanism to manipulate cached children** if we can manipulate cached children may be can solve this problem.

**Comment by flar** [comment]
Yes, a more global strategy may help, but if it is still limited to only transferring at most 3 picture children to the cache per frame, then we have a transition problem if there are 4 children that are uncached as in the first frame where we attempt to transfer the opacity to the children, 3 of them are cached and can do it and the last child was left out of the cache and fails. That last child (actually in the test case there were several children left out on the first cache attempt) flashes when it receives an opacity value and it cannot handle it.
